### PR TITLE
feat: Implement BuildkiteSourceValidator, BuildkiteSourceValidatorFactory

### LINKED ISSUE #1020: Config Validation: Buildkite - implement SourceConfigValidator, SourceConfigValidatorFactory
**Description**
Buildkite - implement SourceConfigValidator, SourceConfigValidatorFactory according to design.

**Acceptance Criteria**
- [ ] BuildkiteSourceConfigValidator implemented
- [ ] BuildkiteSourceConfigValidatorFactory implemented
- [ ] Test BuildkiteSourceConfigValidator and BuildkiteSourceConfigValidatorFactory

**Additional context**
[CI Integrations Config Validator](https://github.com/platform-platform/monorepo/blob/master/metrics/ci_integrations/docs/04_ci_integrations_config_validator.md)

### PR DESCRIPTION
**Pull request by solid-daniyilyevtyushkin**
Fixes #1020

## Checklist

- [x] Documented
- [x] Unit tested

### CODE REVIEW

File: metrics/ci_integrations/test/source/buildkite/config/validator/buildkite_source_validator_test.dart [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+      reset(validationDelegate);
+    });
+
+    PostExpectation<Future<InteractionResult<BuildkiteToken>>>
+        whenValidateAuth() {
+      return when(validationDelegate.validateAuth(auth));
+    }
+
+    PostExpectation<Future<InteractionResult>> whenValidateOrganizationSlug({
+      BuildkiteToken withAccessToken,
```
Comment (solid-vovabeloded) [inline metrics/ci_integrations/test/source/buildkite/config/validator/buildkite_source_validator_test.dart:77]:
```suggestion
      BuildkiteToken accessToken,
```

File: metrics/ci_integrations/test/source/buildkite/config/validator/buildkite_source_validator_test.dart [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+        final result = await validator.validate(config);
+
+        final validationResult = _getAuthValidationResult(result);
+
+        expect(validationResult, MatcherUtil.isSuccessfulValidationResult);
+      },
+    );
+
+    test(
+      ".validate() returns a validation result with the access token field validation result, having the additional context equal to the message returned by the validation delegate if the access token is valid",
```
Comment (solid-vovabeloded) [inline metrics/ci_integrations/test/source/buildkite/config/validator/buildkite_source_validator_test.dart:181]:
```suggestion
      ".validate() returns a validation result with the access token field validation result, having the interaction result message as an additional context",
```

File: metrics/ci_integrations/test/source/buildkite/config/validator/buildkite_source_validator_test.dart [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+
+        final validationResult = _getAuthValidationResult(result);
+        final additionalContext = validationResult.additionalContext;
+
+        expect(additionalContext, equals(message));
+      },
+    );
+
+    test(
+      ".validate() returns a validation result with the unknown organization slug field validation result if the access token is invalid",
```
Comment (solid-vovabeloded) [inline metrics/ci_integrations/test/source/buildkite/config/validator/buildkite_source_validator_test.dart:222]: Would it be better to test that the validate method calls the `setEmptyResults` method after auth token validation failed?

File: metrics/ci_integrations/test/source/buildkite/config/validator/buildkite_source_validator_test.dart [hunk 4]
```diff
@@ -0,0 +1,556 @@
...
+        final expectedAuth = BearerAuthorization(accessToken);
+
+        validator.validate(config);
+
+        verify(validationDelegate.validateAuth(expectedAuth)).called(1);
+      },
+    );
+
+    test(
+      ".validate() returns a validation result with the successful access token field validation result if the access token is valid",
```
Comment (solid-vovabeloded) [inline metrics/ci_integrations/test/source/buildkite/config/validator/buildkite_source_validator_test.dart:168]: Would it be better to test that the validate method sets the right auth field validation result after auth validation?

**Review by solid-vovabeloded — CHANGES_REQUESTED** [review]
Great job, @solid-daniyilyevtyushkin & @solid-yevheniiashchypeleva! I've left a few suggestions below - please take a look.

File: metrics/ci_integrations/test/source/buildkite/config/validator/buildkite_source_validator_test.dart [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+    );
+
+    final validationDelegate = _BuildkiteSourceValidationDelegateMock();
+    final validationResultBuilder = _ValidationResultBuilderMock();
+    final validator = BuildkiteSourceValidator(
+      validationDelegate,
+      validationResultBuilder,
+    );
+
+    tearDown(() {
```
Comment (solid-vovabeloded) [inline metrics/ci_integrations/test/source/buildkite/config/validator/buildkite_source_validator_test.dart:55]: Would it be great to move the `tearDown` method to be right above the test cases?

**Review by solid-vovabeloded — CHANGES_REQUESTED** [review]
Almost there, @solid-daniyilyevtyushkin & @solid-yevheniiashchypeleva! I've added a few more suggestions here - please take a look. Overall, the tests look great, thanks!

**Comment by solid-daniyilyevtyushkin** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by solid-vovabeloded** [comment]
I have read the CLA Document and I hereby sign the CLA

**Review by solid-vovabeloded — APPROVED** [review]
LGTM, thanks @solid-daniyilyevtyushkin & @solid-yevheniiashchypeleva!

**Pull request merged**